// @n3/react-date-picker/es/button.js
import _classCallCheck from "@babel/runtime/helpers/classCallCheck";
import _createClass from "@babel/runtime/helpers/createClass";
import _assertThisInitialized from "@babel/runtime/helpers/assertThisInitialized";
import _inherits from "@babel/runtime/helpers/inherits";
import _createSuper from "@babel/runtime/helpers/createSuper";
import _defineProperty from "@babel/runtime/helpers/defineProperty";
import { PureComponent } from 'react';
import cx from 'classnames';
import { jsx as _jsx } from "react/jsx-runtime";

var Button = /*#__PURE__*/function (_PureComponent) {
  _inherits(Button, _PureComponent);

  var _super = _createSuper(Button);

  function Button() {
    var _this;

    _classCallCheck(this, Button);

    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    _this = _super.call.apply(_super, [this].concat(args));

    _defineProperty(_assertThisInitialized(_this), "onClick", function () {
      var _this$props = _this.props,
          value = _this$props.value,
          onChange = _this$props.onChange;
      onChange(value || null);
    });

    return _this;
  }

  _createClass(Button, [{
    key: "render",
    value: function render() {
      var _this$props2 = this.props,
          className = _this$props2.className,
          disabled = _this$props2.disabled,
          children = _this$props2.children;
      return /*#__PURE__*/_jsx("button", {
        type: "button",
        className: cx('n3__date-picker__button', {
          'n3__date-picker__button_disabled': disabled
        }, className),
        onClick: this.onClick,
        disabled: disabled,
        children: children
      });
    }
  }]);

  return Button;
}(PureComponent);

_defineProperty(Button, "defaultProps", {
  disabled: false,
  className: null,
  value: null
});

export default Button;
