// @n3/react-date-picker/es/field.js
import _objectSpread from "@babel/runtime/helpers/objectSpread2";
import _slicedToArray from "@babel/runtime/helpers/slicedToArray";
import { memo, useCallback } from 'react';
import { useDropdownToggle } from 'react-overlays/DropdownToggle';
import { jsx as _jsx } from "react/jsx-runtime";
import { Fragment as _Fragment } from "react/jsx-runtime";
import { jsxs as _jsxs } from "react/jsx-runtime";

function FieldInner(_ref) {
  var range = _ref.range,
      startDate = _ref.startDate,
      endDate = _ref.endDate,
      canClear = _ref.canClear,
      displayFormat = _ref.displayFormat,
      locale = _ref.locale,
      onClear = _ref.onClear;
  var onClearHandler = useCallback(function (event) {
    event.stopPropagation();
    onClear(event);
  }, [onClear]);

  var _useDropdownToggle = useDropdownToggle(),
      _useDropdownToggle2 = _slicedToArray(_useDropdownToggle, 2),
      toggleProps = _useDropdownToggle2[0],
      _useDropdownToggle2$ = _useDropdownToggle2[1],
      toggle = _useDropdownToggle2$.toggle,
      show = _useDropdownToggle2$.show;

  var onClick = useCallback(function (event) {
    toggle(!show, event);
  }, [show, toggle]);
  return /*#__PURE__*/_jsxs("div", _objectSpread(_objectSpread({
    role: "button",
    className: "n3__date-picker-field",
    onKeyDown: onClick,
    tabIndex: 0
  }, toggleProps), {}, {
    children: [/*#__PURE__*/_jsxs("div", {
      className: "n3__date-picker-field__labels",
      children: [startDate && /*#__PURE__*/_jsx("div", {
        className: "n3__date-picker-field__label",
        children: displayFormat(startDate, locale)
      }), range && endDate && /*#__PURE__*/_jsxs(_Fragment, {
        children: [/*#__PURE__*/_jsx("div", {
          className: "n3__date-picker-field__dash",
          children: "\u2013"
        }), /*#__PURE__*/_jsx("div", {
          className: "n3__date-picker-field__label",
          children: displayFormat(endDate, locale)
        })]
      })]
    }), canClear && /*#__PURE__*/_jsx("button", {
      type: "button",
      className: "n3__date-picker-field__clear-button",
      onClick: onClearHandler,
      "aria-label": "clear button"
    }), /*#__PURE__*/_jsx("div", {
      className: "n3__date-picker-field__calendar"
    })]
  }));
}

FieldInner.defaultProps = {
  startDate: null,
  endDate: null
};
export default /*#__PURE__*/memo(FieldInner);
