// @n3/react-date-picker/es/month.js
import { memo } from 'react';
import cx from 'classnames';
import locales from "./locales";
import Button from "./button";
import { jsx as _jsx } from "react/jsx-runtime";
export var defaultRenderLabel = function defaultRenderLabel(date, locale) {
  return locales[locale].monthsShort[date.getMonth()];
};
export function MonthComponent(_ref) {
  var locale = _ref.locale,
      renderLabel = _ref.renderLabel,
      isCurrent = _ref.isCurrent,
      isDisabled = _ref.isDisabled,
      isSelected = _ref.isSelected,
      value = _ref.value,
      onChange = _ref.onChange;
  return /*#__PURE__*/_jsx(Button, {
    className: cx('n3__date-picker__month', {
      'n3__date-picker__month_current': isCurrent,
      'n3__date-picker__month_selected': isSelected,
      'n3__date-picker__month_disabled': isDisabled
    }),
    disabled: isDisabled,
    value: value,
    onChange: onChange,
    children: renderLabel && renderLabel(value, locale)
  });
}
MonthComponent.defaultProps = {
  renderLabel: defaultRenderLabel
};
export default /*#__PURE__*/memo(MonthComponent);
