// @n3/react-date-picker/es/scrollArea.js
import _classCallCheck from "@babel/runtime/helpers/classCallCheck";
import _createClass from "@babel/runtime/helpers/createClass";
import _assertThisInitialized from "@babel/runtime/helpers/assertThisInitialized";
import _inherits from "@babel/runtime/helpers/inherits";
import _createSuper from "@babel/runtime/helpers/createSuper";
import _defineProperty from "@babel/runtime/helpers/defineProperty";
import { PureComponent } from 'react';
import getScrollLineHeight from "./utils/getScrollLineHeight";
import scheduleRAF from "./utils/scheduleRaf";
import { jsx as _jsx } from "react/jsx-runtime";

var ScrollArea = /*#__PURE__*/function (_PureComponent) {
  _inherits(ScrollArea, _PureComponent);

  var _super = _createSuper(ScrollArea);

  function ScrollArea() {
    var _this;

    _classCallCheck(this, ScrollArea);

    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    _this = _super.call.apply(_super, [this].concat(args));

    _defineProperty(_assertThisInitialized(_this), "deltaY", 0);

    _defineProperty(_assertThisInitialized(_this), "clientY", 0);

    _defineProperty(_assertThisInitialized(_this), "scrollLineHeight", getScrollLineHeight());

    _defineProperty(_assertThisInitialized(_this), "onWheel", function (_ref) {
      var deltaY = _ref.deltaY,
          deltaMode = _ref.deltaMode;

      if (_this.scrollLineHeight === null) {
        return;
      }

      var deltaYValue = deltaMode === WheelEvent.DOM_DELTA_LINE ? deltaY * _this.scrollLineHeight : deltaY;
      _this.deltaY += deltaYValue;

      _this.handleScroll();
    });

    _defineProperty(_assertThisInitialized(_this), "onTouchStart", function (_ref2) {
      var targetTouches = _ref2.targetTouches;
      _this.clientY = targetTouches[0].clientY;
    });

    _defineProperty(_assertThisInitialized(_this), "onTouchMove", function (_ref3) {
      var targetTouches = _ref3.targetTouches;
      var clientY = targetTouches[0].clientY;
      _this.deltaY -= clientY - _this.clientY;
      _this.clientY = clientY;

      _this.handleScroll();
    });

    _defineProperty(_assertThisInitialized(_this), "onTouchEnd", function () {
      _this.clientY = 0;
    });

    _defineProperty(_assertThisInitialized(_this), "handleScroll", scheduleRAF(function () {
      var onChange = _this.props.onChange;
      onChange(_this.deltaY);
      _this.deltaY = 0;
    }));

    return _this;
  }

  _createClass(ScrollArea, [{
    key: "render",
    value: function render() {
      var _this$props = this.props,
          style = _this$props.style,
          className = _this$props.className,
          children = _this$props.children;
      return /*#__PURE__*/_jsx("div", {
        style: style,
        className: className,
        onWheel: this.onWheel,
        onTouchStart: this.onTouchStart,
        onTouchMove: this.onTouchMove,
        onTouchEnd: this.onTouchEnd,
        children: children
      });
    }
  }]);

  return ScrollArea;
}(PureComponent);

_defineProperty(ScrollArea, "defaultProps", {
  style: {},
  className: null
});

export default ScrollArea;
