// @n3/react-date-picker/es/years.js
import _classCallCheck from "@babel/runtime/helpers/classCallCheck";
import _createClass from "@babel/runtime/helpers/createClass";
import _assertThisInitialized from "@babel/runtime/helpers/assertThisInitialized";
import _inherits from "@babel/runtime/helpers/inherits";
import _createSuper from "@babel/runtime/helpers/createSuper";
import _defineProperty from "@babel/runtime/helpers/defineProperty";
import { PureComponent } from 'react';
import addYears from 'date-fns/addYears';
import isSameYear from 'date-fns/isSameYear';
import subYears from 'date-fns/subYears';
import defaultRenderLabel from "./defaults/renderYearLabel";
import ScrollArea from "./scrollArea";
import Year from "./year";
import getDateInterval from "./utils/getDateInterval";
import { jsx as _jsx } from "react/jsx-runtime";

var Years = /*#__PURE__*/function (_PureComponent) {
  _inherits(Years, _PureComponent);

  var _super = _createSuper(Years);

  function Years() {
    var _this;

    _classCallCheck(this, Years);

    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    _this = _super.call.apply(_super, [this].concat(args));

    _defineProperty(_assertThisInitialized(_this), "handleChange", function (deltaY) {
      var _this$props = _this.props,
          selected = _this$props.selected,
          onScroll = _this$props.onScroll;

      if (!selected) {
        return;
      }

      var nextSelected = deltaY > 0 ? subYears(selected, 1) : addYears(selected, 1);
      onScroll(nextSelected);
    });

    return _this;
  }

  _createClass(Years, [{
    key: "render",
    value: function render() {
      var _this$props2 = this.props,
          selected = _this$props2.selected,
          isDisabled = _this$props2.isDisabled,
          onScroll = _this$props2.onScroll,
          renderLabel = _this$props2.renderLabel,
          locale = _this$props2.locale;
      var years = selected ? getDateInterval(addYears(selected, -5), 12) : [];
      return /*#__PURE__*/_jsx(ScrollArea, {
        className: "n3__date-picker__years",
        onChange: this.handleChange,
        children: years.map(function (value, i) {
          return /*#__PURE__*/_jsx(Year, {
            value: value,
            isCurrent: isSameYear(value, new Date()),
            isDisabled: isDisabled(value, isSameYear),
            isSelected: selected ? isSameYear(selected, value) : false,
            onChange: onScroll,
            children: renderLabel && renderLabel(value, locale)
          }, i);
        })
      });
    }
  }]);

  return Years;
}(PureComponent);

_defineProperty(Years, "defaultProps", {
  renderLabel: defaultRenderLabel
});

export default Years;
