// @n3/react-date-picker/lib/calendar.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault")["default"];

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = void 0;

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _assertThisInitialized2 = _interopRequireDefault(require("@babel/runtime/helpers/assertThisInitialized"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _createSuper2 = _interopRequireDefault(require("@babel/runtime/helpers/createSuper"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _react = require("react");

var _classnames = _interopRequireDefault(require("classnames"));

var _locales = _interopRequireDefault(require("./locales"));

var _days = _interopRequireDefault(require("./days"));

var _months = _interopRequireDefault(require("./months"));

var _years = _interopRequireDefault(require("./years"));

var _getCssValue = _interopRequireDefault(require("./utils/getCssValue"));

var _jsxRuntime = require("react/jsx-runtime");

var isWeekend = function isWeekend(weekdayIndex) {
  return [0, 6].includes(weekdayIndex);
};

var preventDefault = function preventDefault(event) {
  event.preventDefault();
};

var Calendar = /*#__PURE__*/function (_Component) {
  (0, _inherits2["default"])(Calendar, _Component);

  var _super = (0, _createSuper2["default"])(Calendar);

  function Calendar(props) {
    var _this;

    (0, _classCallCheck2["default"])(this, Calendar);
    _this = _super.call(this, props);
    (0, _defineProperty2["default"])((0, _assertThisInitialized2["default"])(_this), "calendarHeight", void 0);
    (0, _defineProperty2["default"])((0, _assertThisInitialized2["default"])(_this), "monthsSelectionOffset", void 0);
    (0, _defineProperty2["default"])((0, _assertThisInitialized2["default"])(_this), "monthsSelectionSpace", void 0);
    (0, _defineProperty2["default"])((0, _assertThisInitialized2["default"])(_this), "datesRef", /*#__PURE__*/(0, _react.createRef)());
    _this.calendarHeight = (0, _getCssValue["default"])('--calendar-height');
    _this.monthsSelectionOffset = (0, _getCssValue["default"])('--months-selection-height') / 2;
    _this.monthsSelectionSpace = (0, _getCssValue["default"])('--months-space');
    return _this;
  }

  (0, _createClass2["default"])(Calendar, [{
    key: "componentDidMount",
    value: function componentDidMount() {
      var datesRef = this.datesRef.current;

      if (datesRef) {
        datesRef.addEventListener('wheel', preventDefault);
        datesRef.addEventListener('touchmove', preventDefault);
      }
    }
  }, {
    key: "componentWillUnmount",
    value: function componentWillUnmount() {
      var datesRef = this.datesRef.current;

      if (datesRef) {
        datesRef.removeEventListener('wheel', preventDefault);
        datesRef.removeEventListener('touchmove', preventDefault);
      }
    }
  }, {
    key: "render",
    value: function render() {
      var _this$props = this.props,
          locale = _this$props.locale,
          selected = _this$props.selected,
          onChange = _this$props.onChange,
          onDayChange = _this$props.onDayChange,
          isDisabled = _this$props.isDisabled,
          isBordered = _this$props.isBordered,
          isSelected = _this$props.isSelected;
      var calendarHeight = this.calendarHeight,
          monthsSelectionSpace = this.monthsSelectionSpace,
          monthsSelectionOffset = this.monthsSelectionOffset;
      var _locales$locale = _locales["default"][locale],
          weekdaysOrdering = _locales$locale.weekdaysOrdering,
          weekdays = _locales$locale.weekdays;
      return /*#__PURE__*/(0, _jsxRuntime.jsxs)("div", {
        className: "n3__date-picker-calendar",
        children: [/*#__PURE__*/(0, _jsxRuntime.jsx)("div", {
          className: "n3__date-picker-calendar__weekdays",
          children: weekdaysOrdering.map(function (weekdayIndex) {
            return /*#__PURE__*/(0, _jsxRuntime.jsx)("span", {
              className: (0, _classnames["default"])('n3__date-picker-calendar__weekday', {
                'n3__date-picker-calendar__weekday_weekend': isWeekend(weekdayIndex)
              }),
              children: weekdays[weekdayIndex]
            }, weekdayIndex);
          })
        }), /*#__PURE__*/(0, _jsxRuntime.jsxs)("div", {
          ref: this.datesRef,
          className: "n3__date-picker-calendar__dates",
          children: [/*#__PURE__*/(0, _jsxRuntime.jsx)(_days["default"], {
            locale: locale,
            selected: selected,
            isSelected: isSelected,
            isBordered: isBordered,
            isDisabled: isDisabled,
            onDayChange: onDayChange,
            onScroll: onChange,
            calendarHeight: calendarHeight
          }), /*#__PURE__*/(0, _jsxRuntime.jsx)(_months["default"], {
            locale: locale,
            selected: selected,
            format: "MMM",
            onChange: onChange,
            onScroll: onChange,
            isDisabled: isDisabled,
            calendarHeight: calendarHeight,
            offset: monthsSelectionOffset,
            space: monthsSelectionSpace
          }), /*#__PURE__*/(0, _jsxRuntime.jsx)(_years["default"], {
            locale: locale,
            selected: selected,
            onScroll: onChange,
            isDisabled: isDisabled
          })]
        })]
      });
    }
  }]);
  return Calendar;
}(_react.Component);

var _default = Calendar;
exports["default"] = _default;
