// @n3/react-date-picker/lib/field.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault")["default"];

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = void 0;

var _objectSpread2 = _interopRequireDefault(require("@babel/runtime/helpers/objectSpread2"));

var _slicedToArray2 = _interopRequireDefault(require("@babel/runtime/helpers/slicedToArray"));

var _react = require("react");

var _DropdownToggle = require("react-overlays/DropdownToggle");

var _jsxRuntime = require("react/jsx-runtime");

function FieldInner(_ref) {
  var range = _ref.range,
      startDate = _ref.startDate,
      endDate = _ref.endDate,
      canClear = _ref.canClear,
      displayFormat = _ref.displayFormat,
      locale = _ref.locale,
      onClear = _ref.onClear;
  var onClearHandler = (0, _react.useCallback)(function (event) {
    event.stopPropagation();
    onClear(event);
  }, [onClear]);

  var _useDropdownToggle = (0, _DropdownToggle.useDropdownToggle)(),
      _useDropdownToggle2 = (0, _slicedToArray2["default"])(_useDropdownToggle, 2),
      toggleProps = _useDropdownToggle2[0],
      _useDropdownToggle2$ = _useDropdownToggle2[1],
      toggle = _useDropdownToggle2$.toggle,
      show = _useDropdownToggle2$.show;

  var onClick = (0, _react.useCallback)(function (event) {
    toggle(!show, event);
  }, [show, toggle]);
  return /*#__PURE__*/(0, _jsxRuntime.jsxs)("div", (0, _objectSpread2["default"])((0, _objectSpread2["default"])({
    role: "button",
    className: "n3__date-picker-field",
    onKeyDown: onClick,
    tabIndex: 0
  }, toggleProps), {}, {
    children: [/*#__PURE__*/(0, _jsxRuntime.jsxs)("div", {
      className: "n3__date-picker-field__labels",
      children: [startDate && /*#__PURE__*/(0, _jsxRuntime.jsx)("div", {
        className: "n3__date-picker-field__label",
        children: displayFormat(startDate, locale)
      }), range && endDate && /*#__PURE__*/(0, _jsxRuntime.jsxs)(_jsxRuntime.Fragment, {
        children: [/*#__PURE__*/(0, _jsxRuntime.jsx)("div", {
          className: "n3__date-picker-field__dash",
          children: "\u2013"
        }), /*#__PURE__*/(0, _jsxRuntime.jsx)("div", {
          className: "n3__date-picker-field__label",
          children: displayFormat(endDate, locale)
        })]
      })]
    }), canClear && /*#__PURE__*/(0, _jsxRuntime.jsx)("button", {
      type: "button",
      className: "n3__date-picker-field__clear-button",
      onClick: onClearHandler,
      "aria-label": "clear button"
    }), /*#__PURE__*/(0, _jsxRuntime.jsx)("div", {
      className: "n3__date-picker-field__calendar"
    })]
  }));
}

FieldInner.defaultProps = {
  startDate: null,
  endDate: null
};

var _default = /*#__PURE__*/(0, _react.memo)(FieldInner);

exports["default"] = _default;
