// @n3/react-date-picker/lib/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault")["default"];

Object.defineProperty(exports, "__esModule", {
  value: true
});
var _exportNames = {};
exports["default"] = void 0;

var _objectSpread2 = _interopRequireDefault(require("@babel/runtime/helpers/objectSpread2"));

var _regeneratorRuntime2 = _interopRequireDefault(require("@babel/runtime/helpers/regeneratorRuntime"));

var _asyncToGenerator2 = _interopRequireDefault(require("@babel/runtime/helpers/asyncToGenerator"));

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _assertThisInitialized2 = _interopRequireDefault(require("@babel/runtime/helpers/assertThisInitialized"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _createSuper2 = _interopRequireDefault(require("@babel/runtime/helpers/createSuper"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _react = require("react");

var _Dropdown = _interopRequireDefault(require("react-overlays/Dropdown"));

var _classnames = _interopRequireDefault(require("classnames"));

var _field = _interopRequireDefault(require("./field"));

var _menu = _interopRequireDefault(require("./menu"));

var _locales = require("./locales");

var _parseDate = _interopRequireDefault(require("./defaults/parseDate"));

var _displayFormat = _interopRequireDefault(require("./defaults/displayFormat"));

var _jsxRuntime = require("react/jsx-runtime");

var _types = require("./types");

Object.keys(_types).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _types[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function get() {
      return _types[key];
    }
  });
});

var DatePicker = /*#__PURE__*/function (_PureComponent) {
  (0, _inherits2["default"])(DatePicker, _PureComponent);

  var _super = (0, _createSuper2["default"])(DatePicker);

  function DatePicker() {
    var _this;

    (0, _classCallCheck2["default"])(this, DatePicker);

    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    _this = _super.call.apply(_super, [this].concat(args));
    (0, _defineProperty2["default"])((0, _assertThisInitialized2["default"])(_this), "state", {
      show: false
    });
    (0, _defineProperty2["default"])((0, _assertThisInitialized2["default"])(_this), "onToggle", function (isOpened) {
      _this.setState({
        show: isOpened
      });
    });
    (0, _defineProperty2["default"])((0, _assertThisInitialized2["default"])(_this), "onClear", /*#__PURE__*/(0, _asyncToGenerator2["default"])( /*#__PURE__*/(0, _regeneratorRuntime2["default"])().mark(function _callee() {
      var _this$props, range, onChange;

      return (0, _regeneratorRuntime2["default"])().wrap(function _callee$(_context) {
        while (1) {
          switch (_context.prev = _context.next) {
            case 0:
              _this$props = _this.props, range = _this$props.range, onChange = _this$props.onChange;
              _context.next = 3;
              return _this.setState({
                show: false
              });

            case 3:
              _context.next = 5;
              return onChange(range ? {
                from: null,
                to: null
              } : null);

            case 5:
            case "end":
              return _context.stop();
          }
        }
      }, _callee);
    })));
    (0, _defineProperty2["default"])((0, _assertThisInitialized2["default"])(_this), "parseDate", function (value) {
      var _this$props2 = _this.props,
          parseDate = _this$props2.parseDate,
          locale = _this$props2.locale;

      if (!value) {
        return null;
      }

      return parseDate(value, locale);
    });
    (0, _defineProperty2["default"])((0, _assertThisInitialized2["default"])(_this), "parseDateSingle", function (value) {
      var parsed = _this.parseDate(value);

      if (Array.isArray(parsed)) {
        return parsed[0];
      }

      return parsed;
    });
    (0, _defineProperty2["default"])((0, _assertThisInitialized2["default"])(_this), "renderDropdown", function () {
      var _this$props3 = _this.props,
          range = _this$props3.range,
          from = _this$props3.from,
          to = _this$props3.to,
          value = _this$props3.value,
          locale = _this$props3.locale,
          displayFormat = _this$props3.displayFormat,
          className = _this$props3.className,
          minDate = _this$props3.minDate,
          maxDate = _this$props3.maxDate,
          onChange = _this$props3.onChange,
          defaultFocus = _this$props3.defaultFocus,
          popperConfig = _this$props3.popperConfig,
          containerRef = _this$props3.containerRef,
          isDateDisabled = _this$props3.isDateDisabled;
      var startDate = range ? from : value;
      var endDate = range ? to : null;
      var commonProps = {
        startDate: startDate || null,
        endDate: endDate || null,
        range: range || false,
        locale: (0, _locales.hasSupportLocal)(locale || _locales.defaultLocal) && locale || _locales.defaultLocal,
        displayFormat: displayFormat || _displayFormat["default"]
      };
      return /*#__PURE__*/(0, _jsxRuntime.jsxs)("div", {
        className: (0, _classnames["default"])('n3-date-picker', className),
        children: [/*#__PURE__*/(0, _jsxRuntime.jsx)(_field["default"], (0, _objectSpread2["default"])((0, _objectSpread2["default"])({}, commonProps), {}, {
          canClear: _this.canClear(),
          onClear: _this.onClear
        })), /*#__PURE__*/(0, _jsxRuntime.jsx)(_menu["default"], (0, _objectSpread2["default"])((0, _objectSpread2["default"])({}, commonProps), {}, {
          onChange: onChange,
          minDate: minDate,
          maxDate: maxDate,
          defaultFocus: defaultFocus,
          popperConfig: popperConfig,
          parseDate: _this.parseDateSingle,
          containerRef: containerRef,
          isDateDisabled: isDateDisabled
        }))]
      });
    });
    return _this;
  }

  (0, _createClass2["default"])(DatePicker, [{
    key: "canClear",
    value: function canClear() {
      var _this$props4 = this.props,
          clear = _this$props4.clear,
          range = _this$props4.range,
          from = _this$props4.from,
          to = _this$props4.to,
          value = _this$props4.value;

      if (!clear) {
        return false;
      }

      if (range) {
        return Boolean(from) || Boolean(to);
      }

      return Boolean(value);
    }
  }, {
    key: "render",
    value: function render() {
      var show = this.state.show;
      return /*#__PURE__*/(0, _jsxRuntime.jsx)(_Dropdown["default"], {
        drop: "down",
        show: show,
        onToggle: this.onToggle,
        children: this.renderDropdown()
      });
    }
  }]);
  return DatePicker;
}(_react.PureComponent);

(0, _defineProperty2["default"])(DatePicker, "defaultProps", {
  value: null,
  from: null,
  to: null,
  range: false,
  clear: false,
  locale: _locales.defaultLocal,
  parseDate: _parseDate["default"],
  displayFormat: _displayFormat["default"],
  minDate: null,
  maxDate: null,
  className: null,
  defaultFocus: null,
  containerRef: null,
  isDateDisabled: null
});
(0, _defineProperty2["default"])(DatePicker, "displayName", 'ReactDatePicker');
var _default = DatePicker;
exports["default"] = _default;
