// @n3/react-date-picker/lib/input.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault")["default"];

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = void 0;

var _objectSpread2 = _interopRequireDefault(require("@babel/runtime/helpers/objectSpread2"));

var _objectWithoutProperties2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutProperties"));

var _react = require("react");

var _classnames = _interopRequireDefault(require("classnames"));

var _jsxRuntime = require("react/jsx-runtime");

var _excluded = ["name", "hasFocus"];

var Input = function Input(_ref) {
  var name = _ref.name,
      hasFocus = _ref.hasFocus,
      props = (0, _objectWithoutProperties2["default"])(_ref, _excluded);
  var inputRef = (0, _react.useRef)(null);
  var isFocusInput = typeof hasFocus === 'function' ? hasFocus(name) : hasFocus;
  (0, _react.useEffect)(function () {
    var timeoutId = setTimeout(function () {
      var inputNode = inputRef.current;

      if (inputNode) {
        if (isFocusInput) {
          inputNode.focus();
        } else {
          inputNode.blur();
        }
      }
    }, 0);
    return function () {
      clearTimeout(timeoutId);
    };
  }, [isFocusInput]);
  return /*#__PURE__*/(0, _jsxRuntime.jsx)("input", (0, _objectSpread2["default"])({
    ref: inputRef,
    name: name,
    type: "text",
    className: (0, _classnames["default"])('n3__date-picker__input', {
      'n3__date-picker__input_focus': isFocusInput
    })
  }, props));
};

Input.defaultProps = {
  autoComplete: 'off'
};
var _default = Input;
exports["default"] = _default;
