// @n3/react-date-picker/lib/month.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault")["default"];

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.MonthComponent = MonthComponent;
exports.defaultRenderLabel = exports["default"] = void 0;

var _react = require("react");

var _classnames = _interopRequireDefault(require("classnames"));

var _locales = _interopRequireDefault(require("./locales"));

var _button = _interopRequireDefault(require("./button"));

var _jsxRuntime = require("react/jsx-runtime");

var defaultRenderLabel = function defaultRenderLabel(date, locale) {
  return _locales["default"][locale].monthsShort[date.getMonth()];
};

exports.defaultRenderLabel = defaultRenderLabel;

function MonthComponent(_ref) {
  var locale = _ref.locale,
      renderLabel = _ref.renderLabel,
      isCurrent = _ref.isCurrent,
      isDisabled = _ref.isDisabled,
      isSelected = _ref.isSelected,
      value = _ref.value,
      onChange = _ref.onChange;
  return /*#__PURE__*/(0, _jsxRuntime.jsx)(_button["default"], {
    className: (0, _classnames["default"])('n3__date-picker__month', {
      'n3__date-picker__month_current': isCurrent,
      'n3__date-picker__month_selected': isSelected,
      'n3__date-picker__month_disabled': isDisabled
    }),
    disabled: isDisabled,
    value: value,
    onChange: onChange,
    children: renderLabel && renderLabel(value, locale)
  });
}

MonthComponent.defaultProps = {
  renderLabel: defaultRenderLabel
};

var _default = /*#__PURE__*/(0, _react.memo)(MonthComponent);

exports["default"] = _default;
