// @n3/react-date-picker/lib/scrollArea.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault")["default"];

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = void 0;

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _assertThisInitialized2 = _interopRequireDefault(require("@babel/runtime/helpers/assertThisInitialized"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _createSuper2 = _interopRequireDefault(require("@babel/runtime/helpers/createSuper"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _react = require("react");

var _getScrollLineHeight = _interopRequireDefault(require("./utils/getScrollLineHeight"));

var _scheduleRaf = _interopRequireDefault(require("./utils/scheduleRaf"));

var _jsxRuntime = require("react/jsx-runtime");

var ScrollArea = /*#__PURE__*/function (_PureComponent) {
  (0, _inherits2["default"])(ScrollArea, _PureComponent);

  var _super = (0, _createSuper2["default"])(ScrollArea);

  function ScrollArea() {
    var _this;

    (0, _classCallCheck2["default"])(this, ScrollArea);

    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    _this = _super.call.apply(_super, [this].concat(args));
    (0, _defineProperty2["default"])((0, _assertThisInitialized2["default"])(_this), "deltaY", 0);
    (0, _defineProperty2["default"])((0, _assertThisInitialized2["default"])(_this), "clientY", 0);
    (0, _defineProperty2["default"])((0, _assertThisInitialized2["default"])(_this), "scrollLineHeight", (0, _getScrollLineHeight["default"])());
    (0, _defineProperty2["default"])((0, _assertThisInitialized2["default"])(_this), "onWheel", function (_ref) {
      var deltaY = _ref.deltaY,
          deltaMode = _ref.deltaMode;

      if (_this.scrollLineHeight === null) {
        return;
      }

      var deltaYValue = deltaMode === WheelEvent.DOM_DELTA_LINE ? deltaY * _this.scrollLineHeight : deltaY;
      _this.deltaY += deltaYValue;

      _this.handleScroll();
    });
    (0, _defineProperty2["default"])((0, _assertThisInitialized2["default"])(_this), "onTouchStart", function (_ref2) {
      var targetTouches = _ref2.targetTouches;
      _this.clientY = targetTouches[0].clientY;
    });
    (0, _defineProperty2["default"])((0, _assertThisInitialized2["default"])(_this), "onTouchMove", function (_ref3) {
      var targetTouches = _ref3.targetTouches;
      var clientY = targetTouches[0].clientY;
      _this.deltaY -= clientY - _this.clientY;
      _this.clientY = clientY;

      _this.handleScroll();
    });
    (0, _defineProperty2["default"])((0, _assertThisInitialized2["default"])(_this), "onTouchEnd", function () {
      _this.clientY = 0;
    });
    (0, _defineProperty2["default"])((0, _assertThisInitialized2["default"])(_this), "handleScroll", (0, _scheduleRaf["default"])(function () {
      var onChange = _this.props.onChange;
      onChange(_this.deltaY);
      _this.deltaY = 0;
    }));
    return _this;
  }

  (0, _createClass2["default"])(ScrollArea, [{
    key: "render",
    value: function render() {
      var _this$props = this.props,
          style = _this$props.style,
          className = _this$props.className,
          children = _this$props.children;
      return /*#__PURE__*/(0, _jsxRuntime.jsx)("div", {
        style: style,
        className: className,
        onWheel: this.onWheel,
        onTouchStart: this.onTouchStart,
        onTouchMove: this.onTouchMove,
        onTouchEnd: this.onTouchEnd,
        children: children
      });
    }
  }]);
  return ScrollArea;
}(_react.PureComponent);

(0, _defineProperty2["default"])(ScrollArea, "defaultProps", {
  style: {},
  className: null
});
var _default = ScrollArea;
exports["default"] = _default;
