// @n3/react-date-picker/lib/years.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault")["default"];

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = void 0;

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _assertThisInitialized2 = _interopRequireDefault(require("@babel/runtime/helpers/assertThisInitialized"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _createSuper2 = _interopRequireDefault(require("@babel/runtime/helpers/createSuper"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _react = require("react");

var _addYears = _interopRequireDefault(require("date-fns/addYears"));

var _isSameYear = _interopRequireDefault(require("date-fns/isSameYear"));

var _subYears = _interopRequireDefault(require("date-fns/subYears"));

var _renderYearLabel = _interopRequireDefault(require("./defaults/renderYearLabel"));

var _scrollArea = _interopRequireDefault(require("./scrollArea"));

var _year = _interopRequireDefault(require("./year"));

var _getDateInterval = _interopRequireDefault(require("./utils/getDateInterval"));

var _jsxRuntime = require("react/jsx-runtime");

var Years = /*#__PURE__*/function (_PureComponent) {
  (0, _inherits2["default"])(Years, _PureComponent);

  var _super = (0, _createSuper2["default"])(Years);

  function Years() {
    var _this;

    (0, _classCallCheck2["default"])(this, Years);

    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    _this = _super.call.apply(_super, [this].concat(args));
    (0, _defineProperty2["default"])((0, _assertThisInitialized2["default"])(_this), "handleChange", function (deltaY) {
      var _this$props = _this.props,
          selected = _this$props.selected,
          onScroll = _this$props.onScroll;

      if (!selected) {
        return;
      }

      var nextSelected = deltaY > 0 ? (0, _subYears["default"])(selected, 1) : (0, _addYears["default"])(selected, 1);
      onScroll(nextSelected);
    });
    return _this;
  }

  (0, _createClass2["default"])(Years, [{
    key: "render",
    value: function render() {
      var _this$props2 = this.props,
          selected = _this$props2.selected,
          isDisabled = _this$props2.isDisabled,
          onScroll = _this$props2.onScroll,
          renderLabel = _this$props2.renderLabel,
          locale = _this$props2.locale;
      var years = selected ? (0, _getDateInterval["default"])((0, _addYears["default"])(selected, -5), 12) : [];
      return /*#__PURE__*/(0, _jsxRuntime.jsx)(_scrollArea["default"], {
        className: "n3__date-picker__years",
        onChange: this.handleChange,
        children: years.map(function (value, i) {
          return /*#__PURE__*/(0, _jsxRuntime.jsx)(_year["default"], {
            value: value,
            isCurrent: (0, _isSameYear["default"])(value, new Date()),
            isDisabled: isDisabled(value, _isSameYear["default"]),
            isSelected: selected ? (0, _isSameYear["default"])(selected, value) : false,
            onChange: onScroll,
            children: renderLabel && renderLabel(value, locale)
          }, i);
        })
      });
    }
  }]);
  return Years;
}(_react.PureComponent);

(0, _defineProperty2["default"])(Years, "defaultProps", {
  renderLabel: _renderYearLabel["default"]
});
var _default = Years;
exports["default"] = _default;
